Asset URL rewriting in AMP Optimizer for SXG mode

jonluca opened the pull request:
This is the initial work on implementation of https://github.com/ampproject/amp-toolbox/issues/862

I thought I'd put up the diff in progress to clarify some parts of the problem. 

Based on the past conversation:

> ampUrlPrefix is meant for runtime self-hosting. In that case you want to rewrite all script srcs, but not all asset links.

It looks like `packages/optimizer/lib/transformers/RewriteAmpUrls.js` already rewrites scripts and stylesheets, which should be what's required for SXG compatibility.

Will update this diff as this becomes clearer.

sebastianbenz commented:
> It looks like packages/optimizer/lib/transformers/RewriteAmpUrls.js already rewrites scripts and stylesheets, which should be what's required for SXG compatibility.

That's correct. One way to handle this could be to init `ampUrlPrefix` based on the sxg option  in `DomTransformer#setConfig`.

The [OptimizeImages transformer](https://github.com/ampproject/amp-toolbox/blob/main/packages/optimizer/lib/transformers/OptimizeImages.js#L163) would be a good place to add the img URL rewriting.

jonluca commented:
Sounds great! Is there a list of requirements for what other transforms need to be completed? Is rewriting the images to the AMP CDN the last transform? 

Also, just to confirm, the AMP image URLs should be rewritten as such:

`https://i.pinimg.com/236x/b0/7f/1b/b07f1be9f05baad6ec64af0ebeb1cc34.jpg` becomes

`https://i-pinimg-com.cdn.ampproject.org/i/s/i.pinimg.com/236x/b0/7f/1b/b07f1be9f05baad6ec64af0ebeb1cc34.jpg?nii=t`?

As in, we dash-concatenate the domain and subdomain and turn it into a subdomain of the AMP cache, then place the full url as the path in `/i/s/${fullUrl}`?

sebastianbenz commented:
You can use [toolbox-cache-url](https://github.com/ampproject/amp-toolbox/tree/main/packages/cache-url) to calculate the cache origin. The postfix depends on the protocol. Secure connections have the `/s` appended.

* `https => /i/s/` 
* `http  => /i/`

A few other things which might be helpful:

* You can more or less copy the [implementation from the amppackager](https://github.com/ampproject/amppackager/blob/releases/transformer/transformers/urlrewrite.go#L78) in regards to which attributes need to be rewritten.
* [parseSrcSet](packages/optimizer/lib/parseSrcSet) might be useful for rewriting the srcset attribute.
* we use [cssNano](https://github.com/ampproject/amp-toolbox/blob/main/packages/optimizer/lib/transformers/SeparateKeyframes.js#L22) for parsing CSS.

jonluca commented on packages/cli/lib/cmds/optimize.js:
```diff
@@ -26,10 +26,12 @@ class OptimizeCmd {
     const optimized = await this.optimizer_.transformHtml(html, {
       ampUrlPrefix: host,
+      baseDomain: baseDomain,
+      rewriteUrlsToAmpCache: sxgHost,
```
As in the readme, basedomain is the original domain you'd host your content on (for when you pass a relative path to `amp-img`). 

`rewriteUrlsToAmpCache` is going to be the `sxgHost`, i.e. `cdn.ampproject.org`

patrickkettner replied:
The naming is a bit ambiguous - it reads as an action, rather than a label. would `ampCacheURL` fit a bit better, or am I misunderstanding its purpose?

jonluca commented:
Sounds good, thanks for the direction! Really helpful! I've updated the diff to address those, and added some basic tests. I think this is getting pretty close to what we need it to do? 

Should we include the CSS parsing here as well?

nainar commented:
@patrickkettner could you take a look here?

jonluca commented:
@patrickkettner let me know what you think and if anything should be added!

patrickkettner commented on packages/optimizer/README.md:
````diff
@@ -109,6 +111,44 @@ Automatically import any missing AMP Extensions (e.g. amp-carousel).
+<img src="https://example.com/static/img.jpg"/>
+```
+
+#### `fetch`
````
thanks for documenting this!

patrickkettner commented on packages/optimizer/lib/transformers/OptimizeAssets.js:
```diff
@@ -151,28 +152,84 @@ class OptimizeImages {
+    if (!src) {
+      return;
+    }
+    if (src.startsWith('/')) {
```
do you want to support relative urls for parent/sibling directories? e.g. `../../foo.jpg`

patrickkettner commented on packages/optimizer/lib/transformers/OptimizeAssets.js:
```diff
@@ -151,28 +152,84 @@ class OptimizeImages {
+        host,
+        baseDomain
+      );
+    }
```
great work! while you are here, any chance you want to take a crack at background-image, or other css assets?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ampproject/amp-toolbox?pullRequest=884) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ampproject/amp-toolbox?pullRequest=884) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ampproject/amp-toolbox?pullRequest=884) it.</sub>

patrickkettner commented:
@jonluca would you be able to sign the updated CLA?
